### PR TITLE
feat(controller): Support rate-limitng pod creation.

### LINKED ISSUE #4888: Rate-limiting pod creation
Creating 1000s of pods floods and overloads Kubernetes. We should be able to control the rate at which we create resources (and therefore pods).

### PR DESCRIPTION
**Pull request by alexec**
We'll need to experiment with various settings once released.

### CODE REVIEW

File: server/workflow/workflow_server.go [hunk 1]
```diff
@@ -202,8 +203,10 @@ func (s *workflowServer) WatchWorkflows(req *workflowpkg.WatchWorkflowsRequest,
 	}
 	defer watch.Stop()
 
-	log.Debug("Piping events to channel")
-	defer log.Debug("Result channel done")
+	logCtx := log.WithFields(log.Fields{"watchId": rand.Int()})
```
Comment (alexec) [inline server/workflow/workflow_server.go:206]: revert debugging code

File: workflow/controller/pod/log.go [hunk 2]
```diff
@@ -1,16 +0,0 @@
...
-
-	log "github.com/sirupsen/logrus"
-	apiv1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/util/strategicpatch"
-)
-
-func LogChanges(oldPod *apiv1.Pod, newPod *apiv1.Pod) {
-	a, _ := json.Marshal(oldPod)
-	b, _ := json.Marshal(newPod)
-	patch, _ := strategicpatch.CreateTwoWayMergePatch(a, b, &apiv1.Pod{})
```
Comment (alexec) [inline workflow/controller/pod/log.go:14]: this was always returning an error - so this func never worked

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -2121,9 +2152,8 @@ func (woc *wfOperationCtx) markNodeError(nodeName string, err error) *wfv1.NodeS
 
 // markNodePending is a convenience method to mark a node and set the message from the error
 func (woc *wfOperationCtx) markNodePending(nodeName string, err error) *wfv1.NodeStatus {
-	woc.log.Infof("Mark node %s as Pending, due to: %+v", nodeName, err)
-	node := woc.wf.GetNodeByName(nodeName)
-	return woc.markNodePhase(nodeName, wfv1.NodePending, fmt.Sprintf("Pending %s", time.Since(node.StartedAt.Time)))
+	woc.log.Infof("Mark node %s as Pending, due to: %v", nodeName, err)
+	return woc.markNodePhase(nodeName, wfv1.NodePending, err.Error())
```
Comment (alexec) [inline workflow/controller/operator.go:2156]: adding the timestamp to the pending was causing the workflow to be updated when it had not changed and therefore excessive reconciliation

**Comment by alexec** [comment]
`TestArtifactPassing`

**Comment by alexec** [comment]
Do we need this?

**Review by whynowy — APPROVED** [review]

**Pull request merged**